resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/PerformanceCriticalCodeAnalysisStage.cs

```diff
@@ -250,87 +293,78 @@ private static IMethod ExtractMethodDeclarationFromStartCoroutine([NotNull]ICSha
-
-            if (callGraph.TryGetValue(node, out var children))
-            {
-                foreach (var child in children)
-                {
-                    PropagateCostlyMark(child, callGraph, context, visited);
-                }
-            }
-        }
-
-
+        
         // return value only for invocation nodes. If invoked method contain costly method `true` will be returned.
         private class HotMethodAnalyzer : TreeNodeVisitor<HotMethodAnalyzerContext>, IRecursiveElementProcessor<HotMethodAnalyzerContext>
         {
             private readonly IPsiSourceFile mySourceFile;
             private readonly IHighlightingConsumer myConsumer;
+            private readonly Func<bool> myCheckForInterrupt;
             private readonly int myMaxDepth;
 
-            public HotMethodAnalyzer(IPsiSourceFile sourceFile, IHighlightingConsumer consumer, int maxDepth)
+            public HotMethodAnalyzer(IPsiSourceFile sourceFile, IHighlightingConsumer consumer, Func<bool> checkForInterrupt, int maxDepth)
             {
                 mySourceFile = sourceFile;
                 myConsumer = consumer;
+                myCheckForInterrupt = checkForInterrupt;
                 myMaxDepth = maxDepth;
             }
 
+            public override void VisitMethodDeclaration(IMethodDeclaration methodDeclarationParam, HotMethodAnalyzerContext context)
+            {
+                context.MarkCurrentAsVisited();
+            }
+
             public override void VisitInvocationExpression(IInvocationExpression invocationExpressionParam, HotMethodAnalyzerContext context)
             {
                 // Do not add analysis for invocation expression here. Use 'AnalyzeInvocationExpression' for this goal;   
                 AnalyzeInvocationExpression(invocationExpressionParam, context);
                 
                 // Restriction for depth of analysis
                 if (myMaxDepth <= context.Depth)
-                {
                     return;
-                }
 
                 var reference = (invocationExpressionParam.InvokedExpression as IReferenceExpression)?.Reference;
 
-                var declaredElement = reference?.Resolve().DeclaredElement as IMethod;
+                var declaredElement = reference?.Resolve().DeclaredElement;
                 if (declaredElement == null)
                     return;
                 
-                context.RegisterInvocationInMethod(declaredElement, reference);
-                context.MarkCurrentAsVisited();
+                context.RegisterInvocationInMethod(declaredElement, invocationExpressionParam);
 
                 // find all declarations in current file
-                var declarations = declaredElement.GetDeclarationsIn(mySourceFile).Where(t => t.GetSourceFile() == mySourceFile);
+                var declaration = declaredElement.GetDeclarationsIn(mySourceFile).FirstOrDefault(t => t.GetSourceFile() == mySourceFile);
```

Comment:
`find all declarations in current file`
`.FirstOrDefault(...);`

It looks like the comment is outdated.

I wonder why only the first declaration is processed though. It's possible to have multiple declarations of the same method with partial types. E.g.
```cs
partial class P
{
  partial void M();
}

partial class P
{
  partial void M() => Console.WriteLine();
}
```

Current code will process one random possibly empty declaration from the code above.

Reply:
Regarding `GetDeclarationsIn(mySourceFile)` call which returns declarations from other files - it looks like a copy-paste mistake in the source code for `CSharpMethod`. I'd recommend fixing the root problem instead of hacking it multiple times here.